calico/openstack/test/test_plugin_etcd.py

```diff
@@ -738,6 +767,159 @@ def test_not_master_does_not_resync(self):
         self.simulated_time_advance(31)
         self.assertEtcdWrites({})
 
+    def test_not_master_does_not_poll(self):
+        """
+        Test that a driver that is not master does not poll.
+
+        Master would read through etcd db and handle updates
+        """
+        # Initialize the state early to put the elector in place, then override
+        # it to claim that the driver is not master.
+        self.driver._init_state()
+        self.driver.transport.elector.master = lambda *args: False
+
+        self.driver._register_initial_felixes = mock.Mock()
+        self.driver._handle_status_update = mock.Mock()
+
+        # Allow the etcd transport's resync thread to run. Nothing will happen.
+        self.give_way()
+        self.simulated_time_advance(31)
+        self.assertFalse(self.driver._register_initial_felixes.called)
+        self.assertFalse(self.driver._handle_status_update.called)
+
+    def test_status_subtree_read(self):
+        """
+        Test that initial read through etcd status subtree notices felixes.
+        """
+        # Add arbitrary status keys to accumulated etcd db
+        self.etcd_data[FELIX_STATUS_DIR + "/vm_47/status"] = {"time": "2015-08-14T10:37:54"}
+        self.etcd_data[FELIX_STATUS_DIR + "/vm_47/uptime"] = 12
+        self.etcd_data[FELIX_STATUS_DIR + "/machine_123/uptime"] = 17
+        self.etcd_data[FELIX_STATUS_DIR + "/VMachine/status"] = {"time": "1994-06-09T05:13:18"}
+
+        expected_calls = [mock.call(None, TestIfAgentState(host="vm_47",
+                                                           start_flag=True,
+                                                           agent_type=mech_calico.AGENT_TYPE_FELIX,
+                                                           binary='felix-calico-agent')),
+                          mock.call(None, TestIfAgentState(host="machine_123",
+                                                           start_flag=True,
+                                                           agent_type=mech_calico.AGENT_TYPE_FELIX,
+                                                           binary='felix-calico-agent'))]
+
+        self.driver.db = mock.Mock()
+        self.driver.transport = mock.Mock()
+        self.driver.transport.status_client = self.client
+        # Read through status subtree.
+        self.driver._register_initial_felixes()
+        self.driver.db.create_or_update_agent.assert_has_calls(expected_calls,
+                                                               any_order=True)
+
+        # Reset database
+        self.etcd_data = []
```

Comment:
The etcd_data attribute is initialised to a dict, not a list. I'm also not convinced we need to reset at the end of a test; I think this isolation is enforced by nose.